source/code/plugin/filter_health_model_builder.rb

```diff
@@ -0,0 +1,223 @@
+                        # for unknown/none records, update the "monitor state" to be the latest state (new_state) of the monitor instance from the state
+                        signal.state = @state.get_state(signal.monitor_instance_id).new_state
+                    }
+
+                    @generator.update_last_received_records(reduced_records)
+                    all_records = reduced_records.clone
+                    all_records.push(*missing_signals)
+
+                    @log.info "after Adding missing signals all_records.size #{all_records.size}"
+
+                    # build the health model
+                    @model_builder.process_records(all_records)
+                    all_monitors = @model_builder.finalize_model
+
+                    @log.info "after building health_model #{all_monitors.size}"
+
+                    # update the state for aggregate monitors (unit monitors are updated above)
+                    all_monitors.each{|monitor_instance_id, monitor|
+                        if monitor.is_aggregate_monitor
+                            @state.update_state(monitor,
+                                @provider.get_config(monitor.monitor_id)
+                                )
+                        end
+
+                        instance_state = @state.get_state(monitor_instance_id)
+                        #puts "#{monitor_instance_id} #{instance_state.new_state} #{instance_state.old_state} #{instance_state.should_send}"
+                        should_send = instance_state.should_send
+
+                        # always send cluster monitor as a heartbeat
+                        if !should_send && monitor_instance_id != MonitorId::CLUSTER
+                            all_monitors.delete(monitor_instance_id)
+                        end
+                    }
+
+                    @log.info "after optimizing health signals all_monitors.size #{all_monitors.size}"
+
+                    # for each key in monitor.keys,
+                    # get the state from health_monitor_state
+                    # generate the record to send
+                    all_monitors.keys.each{|key|
+                        record = @provider.get_record(all_monitors[key], state)
+                        if record[HealthMonitorRecordFields::MONITOR_ID] == MonitorId::CLUSTER && all_monitors.size > 1
+                            old_state = record[HealthMonitorRecordFields::OLD_STATE]
+                            new_state = record[HealthMonitorRecordFields::NEW_STATE]
+                            if old_state != new_state && @cluster_old_state != old_state && @cluster_new_state != new_state
+                                    ApplicationInsightsUtility.sendCustomEvent("HealthModel_ClusterStateChanged",{"old_state" => old_state , "new_state" => new_state, "monitor_count" => all_monitors.size})
```

Comment:
can u please remove this ? Sending a customevent per state change will be huge amount of data. rather can we track number of state transitions per monitor?

Reply:
2 things. 
1. this is only at the cluster level 
2. to compute number of state transitions per monitor, we need to send a custom event per state change. So i am confused if you are contradicting your own statement :)

Reply:
If a cluster changes state 100 times in an hour, we send 100 events , just for that cluster (and we have ~20k clusters). Rather, can we send a count of number of times the cluster state changed in the past hour for example ?